# Customer Segmentation in a Lisbon Hotel Chain

The objective is to explore the historical customer information of a 4Star Hotel in Lisbon in order to segment customers and discover the distinguishing features of each group.  
  
This should allow market to have a better understanding of customers groups in order to better engage with the customer. These informations may impact several areas of interaction with the customer, eg:
Marketing: channels, timings, reinforcement points, selling points,...
Sales: Pricing , customer value,...
Reception: Types of interaction,...



## DataSet Description

Talk about the variables


## Setup and Import

In [20]:
import os
import csv
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# import category_encoders as ce
import collections
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA

Data Loading and Initial Analysis

In [21]:
ds = pd.read_csv(r'data\dataset.csv', sep=";")
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111733 entries, 0 to 111732
Data columns (total 29 columns):
ID                      111733 non-null int64
Nationality             111733 non-null object
Age                     107561 non-null float64
DaysSinceCreation       111733 non-null int64
NameHash                111733 non-null object
DocIDHash               110732 non-null object
AverageLeadTime         111733 non-null int64
LodgingRevenue          111733 non-null float64
OtherRevenue            111733 non-null float64
BookingsCanceled        111733 non-null int64
BookingsNoShowed        111733 non-null int64
BookingsCheckedIn       111733 non-null int64
PersonsNights           111733 non-null int64
RoomNights              111733 non-null int64
DistributionChannel     111733 non-null object
MarketSegment           111733 non-null object
SRHighFloor             111733 non-null int64
SRLowFloor              111733 non-null int64
SRAccessibleRoom        111733 non-null int64
SRMe

In [22]:
# Display top 10 rows transposed to show all columns
ds.head(10).transpose()

0  \
ID                                                                    1   
Nationality                                                         PRT   
Age                                                                  52   
DaysSinceCreation                                                   440   
NameHash              0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
DocIDHash             0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...   
AverageLeadTime                                                      59   
LodgingRevenue                                                      292   
OtherRevenue                                                       82.3   
BookingsCanceled                                                      1   
BookingsNoShowed                                                      0   
BookingsCheckedIn                                                     2   
PersonsNights                                                         6   
RoomNights                                                            4   
DistributionChannel                                           Corporate   
MarketSegment                                                 Corporate   
SRHighFloor                                                           0   
SRLowFloor                                                            0   
SRAccessibleRoom                                                      0   
SRMediumFloor                                                         0   
SRBathtub                                                             0   
SRShower                                                              0   
SRCrib                                                                0   
SRKingSizeBed                                                         0   
SRTwinBed                                                             0   
SRNearElevator                                                        0   
SRAwayFromElevator                                                    0   
SRNoAlcoholInMiniBar                                                  0   
SRQuietRoom                                                           0   

                                                                      1  \
ID                                                                    2   
Nationality                                                         PRT   
Age                                                                 NaN   
DaysSinceCreation                                                  1385   
NameHash              0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
DocIDHash             0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   
AverageLeadTime                                                      61   
LodgingRevenue                                                      280   
OtherRevenue                                                         53   
BookingsCanceled                                                      0   
BookingsNoShowed                                                      0   
BookingsCheckedIn                                                     1   
PersonsNights                                                        10   
RoomNights                                                            5   
DistributionChannel                               Travel Agent/Operator   
MarketSegment                                     Travel Agent/Operator   
SRHighFloor                                                           0   
SRLowFloor                                                            0   
SRAccessibleRoom                                                      0   
SRMediumFloor                                                         0   
SRBathtub                                                             0   
SRShower                                                              0   
SRCrib                                                                0   
SRKingSizeBed      

In [23]:
# Summary statistics for numerical variables
summary=ds.describe(exclude=[np.object])   # exclude only objects, they will be treated next
summary=summary.transpose()  # transpose the summary for easier reading
summary.head(len(summary))

count          mean           std   min      25%  \
ID                    111733.0  55867.000000  32254.683151   1.0  27934.0   
Age                   107561.0     45.639191     17.244952 -10.0     33.0   
DaysSinceCreation     111733.0    595.026599    374.657382  36.0    288.0   
AverageLeadTime       111733.0     60.833147     85.115320  -1.0      0.0   
LodgingRevenue        111733.0    283.851283    379.131556   0.0      0.0   
OtherRevenue          111733.0     64.682802    123.580715   0.0      0.0   
BookingsCanceled      111733.0      0.002282      0.080631   0.0      0.0   
BookingsNoShowed      111733.0      0.000600      0.028217   0.0      0.0   
BookingsCheckedIn     111733.0      0.737607      0.730889   0.0      0.0   
PersonsNights         111733.0      4.328318      4.630739   0.0      0.0   
RoomNights            111733.0      2.203825      2.301637   0.0      0.0   
SRHighFloor           111733.0      0.042512      0.201755   0.0      0.0   
SRLowFloor            111733.0      0.001307      0.036125   0.0      0.0   
SRAccessibleRoom      111733.0      0.000224      0.014957   0.0      0.0   
SRMediumFloor         111733.0      0.000770      0.027733   0.0      0.0   
SRBathtub             111733.0      0.003132      0.055881   0.0      0.0   
SRShower              111733.0      0.001629      0.040327   0.0      0.0   
SRCrib                111733.0      0.016181      0.126173   0.0      0.0   
SRKingSizeBed         111733.0      0.363268      0.480943   0.0      0.0   
SRTwinBed             111733.0      0.156811      0.363624   0.0      0.0   
SRNearElevator        111733.0      0.000331      0.018195   0.0      0.0   
SRAwayFromElevator    111733.0      0.003598      0.059874   0.0      0.0   
SRNoAlcoholInMiniBar  111733.0      0.000197      0.014031   0.0      0.0   
SRQuietRoom           111733.0      0.087718      0.282886   0.0      0.0   

                          50%      75%        max  
ID                    55867.0  83800.0  111733.00  
Age                      47.0     58.0     123.00  
DaysSinceCreation       522.0    889.0    1385.00  
AverageLeadTime          21.0     95.0     588.00  
LodgingRevenue          208.0    393.3   21781.00  
OtherRevenue             31.0     84.0    8859.25  
BookingsCanceled          0.0      0.0      15.00  
BookingsNoShowed          0.0      0.0       3.00  
BookingsCheckedIn         1.0      1.0      76.00  
PersonsNights             4.0      6.0     116.00  
RoomNights                2.0      3.0     185.00  
SRHighFloor               0.0      0.0       1.00  
SRLowFloor                0.0      0.0       1.00  
SRAccessibleRoom          0.0      0.0       1.00  
SRMediumFloor             0.0      0.0       1.00  
SRBathtub                 0.0      0.0       1.00  
SRShower                  0.0      0.0       1.00  
SRCrib                    0.0      0.0       1.00  
SRKingSizeBed             0.0      1.0       1.00  
SRTwinBed                 0.0      0.0       1.00  
SRNearElevator            0.0      0.0       1.00  
SRAwayFromElevator        0.0      0.0       1.00  
SRNoAlcoholInMiniBar      0.0      0.0       1.00  
SRQuietRoom               0.0      0.0       1.00

In [24]:
# Summary statistics for non numerical variables
summary=ds.describe(include=[np.object],percentiles=None)   
summary=summary.transpose()  
summary.head(len(summary))

count  unique  \
Nationality          111733     199   
NameHash             111733  107584   
DocIDHash            110732  103480   
DistributionChannel  111733       4   
MarketSegment        111733       7   

                                                                   top   freq  
Nationality                                                        FRA  16516  
NameHash             0x15A713CE687991691A18F6CDC56ABE24979C73CF5D51...     75  
DocIDHash            0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   3032  
DistributionChannel                              Travel Agent/Operator  91019  
MarketSegment                                                    Other  63680

199 Nationalities problems probably...
We seem to have duplicates on Name Hashes and DocIDHashes , so we may have to do something about this

In [25]:
nationalities=list(set(ds.Nationality))
print(nationalities)
len(nationalities)

['KOR', 'TGO', 'ECU', 'BHS', 'AIA', 'COM', 'BLR', 'UGA', 'SEN', 'HRV', 'TZA', 'DNK', 'RWA', 'THA', 'AGO', 'PNG', 'ALB', 'SLE', 'DOM', 'CPV', 'CMR', 'NCL', 'PAN', 'CRI', 'KNA', 'TJK', 'JPN', 'DEU', 'DMA', 'ABW', 'LTU', 'VIR', 'EST', 'BEL', 'SOM', 'MAR', 'ASM', 'COK', 'WSM', 'VEN', 'VNM', 'GHA', 'HTI', 'CHN', 'KAZ', 'BMU', 'TMP', 'FIN', 'GIB', 'ISL', 'SUR', 'SWZ', 'CYP', 'ESP', 'UKR', 'RUS', 'KEN', 'GNQ', 'CZE', 'SPM', 'IRL', 'GEO', 'TON', 'BFA', 'AFG', 'NLD', 'MKD', 'GBR', 'ITA', 'TUR', 'PAK', 'DZA', 'MOZ', 'IOT', 'CYM', 'MUS', 'USA', 'IRQ', 'ISR', 'MNE', 'TCD', 'ARG', 'CAN', 'ETH', 'BIH', 'BGD', 'IRN', 'BGR', 'GIN', 'MDV', 'ERI', 'BEN', 'MLI', 'FJI', 'NRU', 'BHR', 'AUS', 'QAT', 'LKA', 'PRY', 'ATA', 'SYC', 'ROU', 'GRC', 'PRI', 'URY', 'LBY', 'MMR', 'YEM', 'VCT', 'FRO', 'JOR', 'MDG', 'KWT', 'GUF', 'PCN', 'NOR', 'PYF', 'LCA', 'LUX', 'GRD', 'TTO', 'ATG', 'FSM', 'MWI', 'GTM', 'KGZ', 'PER', 'HUN', 'LVA', 'SDN', 'NAM', 'BDI', 'SMR', 'BRB', 'SWE', 'LBN', 'ZWE', 'POL', 'FLK', 'PHL', 'EGY', 'NZL'

199

In [26]:
#Check the document hashes
print(ds["DocIDHash"].value_counts(sort=True).head(60))

0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855    3032
0xA486FBACF4B4E5537B026743E3FDFE571D716839E758236F42950A61FE6B922B      31
0x2B17E9D2CCEF2EA0FE752EE345BEDFB06741FFC8ECECF45D6BBDBAF9A274FF52      24
0x469CF1F9CF8C790FFA5AD3F484F2938CBEFF6435BCFD734F687EC6D1E968F076      15
0x2A14D03A4827C67E0D39408F103DB417AD496DCE6158F8309E6281185C042003      14
0x9220D336F2DDD7B68F5066878889C7637EE28924B249F968F5EC82D895B108A7      12
0x3856085146F7BC27BD07BFC4CA1991ED4E65E179D7BDB7DBBA7E32620809C799      12
0xD2DBD6039916F6DB10C6564D8EB9A9116811435965D7D00E7DA292066B3ECE91      11
0x1BF60C4718497A0AB8B46FF00708D3250A484DDA0FDC0248999C782807195BCB      11
0x8FA8EB6D044E4F2C691C2091FAB27B92FEFE22122F419975703C3D5BA76AC4A2      10
0xA89022F442F23A6D7486C47C9F968BF35898B36F0EB3531804CA4613FF33DC45      10
0x10DFBA7DC4CFBBC6403B380AE137098B254DBBCBE5DEEB0C3B240E0F12F0C6D4      10
0x6B421376B94F3D1722979458A96DF486DEA0F9290CC05E9699F2762FD0DDA71D      10
0x1B16B1DF538BA12DC3F97ED

One recurring ID, way too many times to be a person, what to do with others?

In [33]:
from statistics import mode
doc_explore=ds.loc[ds['DocIDHash'] == (mode(ds["DocIDHash"]))]

summary=doc_explore.describe(include="all")
summary=summary.transpose()  
summary.head(len(summary))



#doc_explore.DistributionChannel.value_counts()
#doc_explore.MarketSegment.value_counts()

count unique  \
ID                    3032    NaN   
Nationality           3032      3   
Age                    171    NaN   
DaysSinceCreation     3032    NaN   
NameHash              3032   2826   
DocIDHash             3032      1   
AverageLeadTime       3032    NaN   
LodgingRevenue        3032    NaN   
OtherRevenue          3032    NaN   
BookingsCanceled      3032    NaN   
BookingsNoShowed      3032    NaN   
BookingsCheckedIn     3032    NaN   
PersonsNights         3032    NaN   
RoomNights            3032    NaN   
DistributionChannel   3032      4   
MarketSegment         3032      7   
SRHighFloor           3032    NaN   
SRLowFloor            3032    NaN   
SRAccessibleRoom      3032    NaN   
SRMediumFloor         3032    NaN   
SRBathtub             3032    NaN   
SRShower              3032    NaN   
SRCrib                3032    NaN   
SRKingSizeBed         3032    NaN   
SRTwinBed             3032    NaN   
SRNearElevator        3032    NaN   
SRAwayFromElevator    3032    NaN   
SRNoAlcoholInMiniBar  3032    NaN   
SRQuietRoom           3032    NaN   

                                                                    top  freq  \
ID                                                                  NaN   NaN   
Nationality                                                         PRT  3030   
Age                                                                 NaN   NaN   
DaysSinceCreation                                                   NaN   NaN   
NameHash              0x5175AC9E84362C505AED3E76F20320BE69DD1C21AA67...    10   
DocIDHash             0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...  3032   
AverageLeadTime                                                     NaN   NaN   
LodgingRevenue                                                      NaN   NaN   
OtherRevenue                                                        NaN   NaN   
BookingsCanceled                                                    NaN   NaN   
BookingsNoShowed                                                    NaN   NaN   
BookingsCheckedIn                                                   NaN   NaN   
PersonsNights                                                       NaN   NaN   
RoomNights                                                          NaN   NaN   
DistributionChannel                               Travel Agent/Operator  2138   
MarketSegment                                                     Other  1160   
SRHighFloor                                                         NaN   NaN   
SRLowFloor                                                          NaN   NaN   
SRAccessibleRoom                                                    NaN   NaN   
SRMediumFloor                                                       NaN   NaN   
SRBathtub                                                           NaN   NaN   
SRShower                                                            NaN   NaN   
SRCrib                                                              NaN   NaN   
SRKingSizeBed                                                       NaN   NaN   
SRTwinBed                                                           NaN   NaN   
SRNearElevator                                                      NaN   NaN   
SRAwayFromElevator                                                  NaN   NaN   
SRNoAlcoholInMiniBar                                                NaN   NaN   
SRQuietRoom                                                         NaN   NaN   

                             mean        std  min      25%      50%      75%  \
ID                          27711    24608.2    2  8240.25  20409.5  41602.8   
Nationality                   NaN        NaN  NaN      NaN      NaN      NaN   
Age                        39.924    17.1889    2       29       39     53.5   
DaysSinceCreation         939.059    330.527   38   690.75     1022     1219   
NameHash                      NaN        NaN  NaN      NaN      NaN      NaN   

Maybe an alias for document not presented, or other situation? It has different ages, 2826 different Name Hashes. 

In [34]:
#Check Distribution channel and Market Segment
doc_explore["DistributionChannel","MarketSegment"].apply(pd.Series.value_counts)

KeyError: ('DistributionChannel', 'MarketSegment')

None the less, we must think on the variables that are only available after arrival, since those cannot be used in the model, or predictions cannot be made before arrival.  

And documents are presented on arrival, but... 

In [18]:
doc_explore=ds.loc[ds['DocIDHash'] == "0xA486FBACF4B4E5537B026743E3FDFE571D716839E758236F42950A61FE6B922B"]

summary=doc_explore.describe(include="all")
summary=summary.transpose()
summary.head(len(summary))

count unique  \
ID                      31    NaN   
Nationality             31      1   
Age                     31    NaN   
DaysSinceCreation       31    NaN   
NameHash                31      9   
DocIDHash               31      1   
AverageLeadTime         31    NaN   
LodgingRevenue          31    NaN   
OtherRevenue            31    NaN   
BookingsCanceled        31    NaN   
BookingsNoShowed        31    NaN   
BookingsCheckedIn       31    NaN   
PersonsNights           31    NaN   
RoomNights              31    NaN   
DistributionChannel     31      1   
MarketSegment           31      1   
SRHighFloor             31    NaN   
SRLowFloor              31    NaN   
SRAccessibleRoom        31    NaN   
SRMediumFloor           31    NaN   
SRBathtub               31    NaN   
SRShower                31    NaN   
SRCrib                  31    NaN   
SRKingSizeBed           31    NaN   
SRTwinBed               31    NaN   
SRNearElevator          31    NaN   
SRAwayFromElevator      31    NaN   
SRNoAlcoholInMiniBar    31    NaN   
SRQuietRoom             31    NaN   

                                                                    top freq  \
ID                                                                  NaN  NaN   
Nationality                                                         PRT   31   
Age                                                                 NaN  NaN   
DaysSinceCreation                                                   NaN  NaN   
NameHash              0x8DF2AF984365949E7F4EAB2EBA9BF9CA8DF106B5F2A9...   20   
DocIDHash             0xA486FBACF4B4E5537B026743E3FDFE571D716839E758...   31   
AverageLeadTime                                                     NaN  NaN   
LodgingRevenue                                                      NaN  NaN   
OtherRevenue                                                        NaN  NaN   
BookingsCanceled                                                    NaN  NaN   
BookingsNoShowed                                                    NaN  NaN   
BookingsCheckedIn                                                   NaN  NaN   
PersonsNights                                                       NaN  NaN   
RoomNights                                                          NaN  NaN   
DistributionChannel                                           Corporate   31   
MarketSegment                                                 Corporate   31   
SRHighFloor                                                         NaN  NaN   
SRLowFloor                                                          NaN  NaN   
SRAccessibleRoom                                                    NaN  NaN   
SRMediumFloor                                                       NaN  NaN   
SRBathtub                                                           NaN  NaN   
SRShower                                                            NaN  NaN   
SRCrib                                                              NaN  NaN   
SRKingSizeBed                                                       NaN  NaN   
SRTwinBed                                                           NaN  NaN   
SRNearElevator                                                      NaN  NaN   
SRAwayFromElevator                                                  NaN  NaN   
SRNoAlcoholInMiniBar                                                NaN  NaN   
SRQuietRoom                                                         NaN  NaN   

                          mean       std    min      25%    50%      75%  \
ID                     51674.5     21078  12953  37829.5  48009  63716.5   
Nationality                NaN       NaN    NaN      NaN    NaN      NaN   
Age                         51         0     51       51     51       51   
DaysSinceCreation      602.645   241.352    142      460    588    745.5   
NameHash                   NaN       NaN    NaN      NaN    NaN      NaN   
DocIDHash                  NaN       NaN    NaN      

For this second document hash, we see  some things that are equal across different records, the nationality, the age and we can see more consistency on the values.

In [ ]:
perguntas: some columns are averages, the ones that are binary represent what? The requests of the last booking? What should we do with the records with the same document hash...
    